docs/readers.md

```diff
@@ -0,0 +1,254 @@
+```
+
+Optionally construct with different delimiter, enclosure and/or escape
+character:
+
+```php
+$reader = new CsvReader($file, ';');
+```
+
+Then iterate over the CSV file:
+
+```php
+foreach ($reader as $row) {
+    // $row will be an array containing the comma-separated elements of the line:
+    // array(
+    //   0 => 'James',
+    //   1 => 'Bond'
+    //   etc...
+    // )
+}
+```
+
+### Column headers
+
+If one of your rows contains column headers, you can read them to make the rows
+associative arrays:
+
+```php
+$reader->setHeaderRowNumber(0);
+
+foreach ($reader as $row) {
+    // $row will now be an associative array:
+    // array(
+    //   'firstName' => 'James',
+    //   'lastName'  => 'Bond'
+    //   etc...
+    // )
+}
+```
+
+### Strict mode
+
+The CSV reader operates in strict mode by default. If the reader encounters a
+row where the number of values differs from the number of column headers, an
+error is logged and the row is skipped. Retrieve the errors with `getErrors()`.
+
+To disable strict mode, set `$reader->setStrict(false)` after you instantiate
+the reader.
+
+Disabling strict mode means:
+
+1. Any rows that contain fewer values than the column headers are simply
+   padded with null values.
+2. Any additional values in a row that contain more values than the
+   column headers are ignored.
+
+Examples where this is useful:
+
+- **Outlook 2010:** which omits trailing blank values
+- **Google Contacts:** which exports more values than there are column headers
+
+### Duplicate headers
+
+Sometimes a CSV file contains duplicate column headers, for instance:
+
+id  | details  | details
+--- | -------- | --------
+1   | bla      | more bla
+
+By default, a `DuplicateHeadersException` will be thrown if you call
+`setHeaderRowNumber(0)` on this file. You can handle duplicate columns in
+one of three ways:
+* call `setColumnHeaders(['id', 'details', 'details_2'])` to specify your own
+  headers
+* call `setHeaderRowNumber` with the `CsvReader::DUPLICATE_HEADERS_INCREMENT`
+  flag to generate incremented headers; in this case: `id`, `details` and
+  `details1`
+* call `setHeaderRowNumber` with the `CsvReader::DUPLICATE_HEADERS_MERGE` flag
+  to merge duplicate values into arrays; in this case, the first row’s values
+  will become: `[ 'id' => 1, 'details' => [ 'bla', 'more bla' ] ]`.
+
+## Doctrine DBAL
+
+Reads data through [Doctrine’s DBAL](http://www.doctrine-project.org/projects/dbal.html).
+Your project should include Doctrine’s DBAL package:
+
+```bash
+$ composer require doctrine/dbal
+```
+
+```php
+use Port\Reader\Dbal\DbalReader;
+
+$reader = new DbalReader(
+    $connection, // Instance of \Doctrine\DBAL\Connection
+    'SELECT u.id, u.username, g.name FROM `user` u INNER JOIN groups g ON u.group_id = g.id'
+);
+```
+
+## Doctrine ORM/ODM
+
+Reads data through the [Doctrine ORM](http://www.doctrine-project.org/projects/orm.html):
+
+```php
+use Port\Reader\Doctrine\DoctrineReader;
```

Comment:
I am pretty sure it is almost impossible to rebase that branch so it can work. :wink: